doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,514 @@
+2. User could use SONiC CLI command "config apply-patch" to incrementally update configurations systematically and validate target configuration using SONiC Yang models. With generic_config_updater, all operations in a 'config apply-patch' command are processed in a single transaction that will either succeed or fail as a whole.
+    - "config apply-patch" supports ConfigDB, and Yang models.
+    - The green arrow indicates this process, and dotted arrow indicates Yang validation.
+3. SONiC CLI commands can use "config reload" to fully update configurations and validate target configuration using SONiC Yang models.
+    - "config reload" already support input in ConfigDB schema or in yang models schema.
+    - The yellow arrow indicates this process, and dotted arrow indicates Yang validation.
+
+And we have the possibility to improve generic_config_updater performance.
+
+* SONIC Restapi
+
+<img src="images/restapi.png" alt="restapi" width="500px"/>
+
+sonic-restapi is a docker container running in SONiC, which exposes HTTPS endpoints to perform dynamic configuration including config VNET routes.
+One challenge of sonic-restapi is to provide a general config schema. Currently its API is designed case by case.
+
+sonic-restapi will be replaced by gNMI interface in the future.
+
+* SONiC system telemetry services
+
+<img src="images/telemetry.png" alt="telemetry" width="500px"/>
+
+sonic-telemetry provides gNMI/gNOI server interface. And sonic-telemetry provides only telemetry via gRPC, and the gap is configuration.
+
+## 1.1 Requirements
+
+* Set and get RPCs must be supported. Customer will use get RPC to retrieve configurations including FRR and VNET route, and use set PRC to apply new configurations.
+* SetRequest message must support to incrementally update configuration and fully update configurations.
+* Data models can be SONiC Yang models or CONFIG_DB schema.
+* Configurations must be verified using YANG models even it is in CONFIG_DB schema.
+* Need to configure huge VNET route entries to ApplDB, with high speed.
+* Configurations must be persisted after a device restart.
+* Need to support multi-ASIC platform.
+* Support authentication and authorization with TACACS+ server.
+* [low-priority] FRR Yang models will also be supported.
+* [low-priority] Upgrade operation will be provided with GNOI.
+* Must follow gnmi-specification.
+    * All changes to the state of the target that are included in an individual SetRequest message are considered part of a transaction.
+    * Within an individual transaction (SetRequest) the order of operations is 'delete', 'replace', 'update'.
+    * If a single path is in multiple operations in one SetRequest (i.e., within 'update' or 'replace'), then the state of the target MUST reflect the application of all the operations in order, even if they overwrite each other.
+    * The session between the client and server MUST be encrypted using TLS.
+    * New connections are mutually authenticated.
```

Comment:
I also suggest adding a section for heart-beat mechanism and reset-status of the gnmi container. For e.g, if the gnmi container restarts, does clients have to reprogram or not? Full system reload vs just gnmi container reload.

Reply:
Heart-beat mechanism is used to detect reload?
If gnmi container restarts, client doesn't need to reprogram, but for system reload, client needs to reprogram.
I think client can use heart-beat to detect reload, and then use gnmi get to check if needs to reprogram.

Reply:
Updated.